xyz

```diff
@@ -60,27 +65,70 @@ Options:
         Print xyz's version number and exit.
 "
 
-# http://stackoverflow.com/a/246128/312785
-unset CDPATH
-path="${BASH_SOURCE[0]}"
-while [ -h "$path" ] ; do
-  dir="$(cd -P "$(dirname "$path")" && pwd)"
-  path="$(readlink "$path")"
-  [[ $path == /* ]] || path="$dir/$path"
-done
-dir="$(cd -P "$(dirname "$path")" && pwd)"
+inc() {
+  local prerelease_label="$1" increment="$2" version="$3"
+  local number='0|[1-9][0-9]*'
+  local part="$number|[-0-9A-Za-z]+"
+  local pattern="^($number)[.]($number)[.]($number)(-(($part)([.]($part))*))?$"
+  if ! [[ $version =~ $pattern ]] ; then
+    printf 'Invalid version: %s\n' "$version" >&2
+    return 1
+  fi
+  local -i x=${BASH_REMATCH[1]}
+  local -i y=${BASH_REMATCH[2]}
+  local -i z=${BASH_REMATCH[3]}
+  local suff=${BASH_REMATCH[5]}
+
+  local -a parts=("$prerelease_label" 0)
+  case $increment in
+    major)  (( y + z == 0 )) && [[ -n $suff ]] || x+=1 ; y=0 ; z=0 ;;
+    minor)      (( z == 0 )) && [[ -n $suff ]] || y+=1 ; z=0 ;;
+    patch)                      [[ -n $suff ]] || z+=1 ;;
+    premajor)                                     x+=1 ; y=0 ; z=0 ;;
+    preminor)                                     y+=1 ; z=0 ;;
+    prepatch)                                     z+=1 ;;
```

Comment:
The majority of the algorithm is implemented in the six lines above. It reads better to me than the mutation-laden JavaScript implementation. ;)